contracts/core/BorrowableMarginPool.sol

```diff
@@ -0,0 +1,288 @@
+        return whitelistedBorrower[_borrower];
+    }
+
+    /**
+     * @notice check if a oToken buyer is whitelisted
+     * @param _oTokenBuyer address of the oToken buyer
+     * @return boolean, True if the oToken buyer is whitelisted
+     */
+    function isWhitelistedOTokenBuyer(address _oTokenBuyer) external view returns (bool) {
+        return whitelistedOTokenBuyer[_oTokenBuyer];
+    }
+
+    /**
+     * @notice check if a options vault is whitelisted
+     * @param _optionsVault address of the options vault
+     * @return boolean, True if the options vault is whitelisted
+     */
+    function isWhitelistedOptionsVault(address _optionsVault) external view returns (bool) {
+        return whitelistedOptionsVault[_optionsVault];
+    }
+
+    /**
+     * @notice Set borrower whitelist status
+     * @param _borrower address of the borrower (100% of the time verified market makers)
+     * @param _whitelisted bool of whether it is whitelisted or blacklisted
+     */
+    function setBorrowerWhitelistedStatus(address _borrower, bool _whitelisted) external onlyOwner {
+        require(_borrower != address(0), "!_borrower");
+
+        whitelistedBorrower[_borrower] = _whitelisted;
+        emit SetBorrowWhitelist(_borrower, _whitelisted);
+    }
+
+    /**
+     * @notice Set options vault whitelist status
+     * @param _optionsVault address of the options vault
+     * @param _whitelisted bool of whether it is whitelisted or blacklisted
+     */
+    function setOptionsVaultWhitelistedStatus(address _optionsVault, bool _whitelisted) external onlyOwner {
+        require(_optionsVault != address(0), "!_optionsVault");
+
+        whitelistedOptionsVault[_optionsVault] = _whitelisted;
+        emit SetOptionsVaultWhitelist(_optionsVault, _whitelisted);
+    }
+
+    /**
+     * @notice Set oToken buyer whitelist status
+     * @param _oTokenBuyer address of the oToken buyer
+     * @param _whitelisted bool of whether it is whitelisted or blacklisted
+     */
+    function setOTokenBuyerWhitelistedStatus(address _oTokenBuyer, bool _whitelisted) external onlyOwner {
+        require(_oTokenBuyer != address(0), "!_buyer");
+
+        whitelistedOTokenBuyer[_oTokenBuyer] = _whitelisted;
+        emit SetOTokenBuyerWhitelist(_oTokenBuyer, _whitelisted);
+    }
+
+    /**
+     * @notice Set Borrow percent of collateral asset
+     * @param _collateral address of collateral asset
+     * @param _borrowPCT borrow PCT
+     */
+    function setBorrowPCT(address _collateral, uint256 _borrowPCT) external onlyOwner {
+        borrowPCT[_collateral] = _borrowPCT;
+        emit SetBorrowPCT(_collateral, _borrowPCT);
+    }
+
+    /**
+     * @notice Lends out asset to market maker
+     * @param _oToken address of the oToken laying claim to the collateral assets
+     * @param _amount amount of the asset to lend to borrower
+     */
+    function borrow(address _oToken, uint256 _amount) public onlyWhitelistedBorrower {
```

Comment:
I guess here it make more sense to allow borrower to set the amount of oToken they are giving, and calculate amount of collateral asset to borrow, instead of doing it the other way, also allow for borrower to borrow and keep some oToken as this is using all the msg.sender oToken balance.